crates/rethnet_eth/src/remote/client.rs

```diff
@@ -0,0 +1,777 @@
+use std::{
+    io,
+    sync::atomic::{AtomicU64, Ordering},
+};
+
+use itertools::Itertools;
+use revm_primitives::{AccountInfo, Address, Bytecode, B256, U256};
+
+use super::{eth, jsonrpc, BlockSpec, GetLogsInput, MethodInvocation, SerializableBlockSpec};
+
+/// Specialized error types
+#[derive(thiserror::Error, Debug)]
+pub enum RpcClientError {
+    /// The remote node's response did not conform to the expected format
+    #[error("Response body `{body}` did not contain the expected type: `{expected_type}`: {msg}")]
+    InterpretationError {
+        /// The error message
+        msg: String,
+        /// The Rust type which was expected to be decoded from the JSON received
+        expected_type: &'static str,
+        /// The body of the response given by the remote node
+        body: String,
+    },
+
+    /// The message could not be sent to the remote node
+    #[error(transparent)]
+    FailedToSend(reqwest::Error),
+
+    /// The remote node failed to reply with the body of the response
+    #[error("The response text was corrupted: {0}.")]
+    CorruptedResponse(reqwest::Error),
+
+    /// The JSON-RPC returned an error.
+    #[error("{error}. Request: {request}")]
+    JsonRpcError {
+        /// The JSON-RPC error
+        error: jsonrpc::Error,
+        /// The request JSON
+        request: String,
+    },
+
+    /// Some other error from an underlying dependency
+    #[error(transparent)]
+    OtherError(#[from] io::Error),
+}
+
+#[derive(serde::Serialize)]
+struct Request<'a> {
+    version: jsonrpc::Version,
+    #[serde(flatten)]
+    method: &'a MethodInvocation,
+    id: &'a jsonrpc::Id,
+}
+
+struct BatchResponse {
+    text: String,
+    request_ids: Vec<jsonrpc::Id>,
+    request_strings: Vec<String>,
+}
+
+/// A client for executing RPC methods on a remote Ethereum node
+#[derive(Debug)]
+pub struct RpcClient {
+    url: String,
+    client: reqwest::Client,
+    next_id: AtomicU64,
+}
+
+impl RpcClient {
+    fn extract_response<T>(response: &str, request_id: &jsonrpc::Id) -> Result<T, jsonrpc::Error>
+    where
+        T: for<'a> serde::Deserialize<'a>,
+    {
+        let response: jsonrpc::Response<T> =
+            serde_json::from_str(response).unwrap_or_else(|error| {
+                panic!(
+                    "Response `{response}` failed to parse with expected type `{expected_type}`, due to error: {error}",
+                    expected_type = std::any::type_name::<T>()
+                )
```

Comment:
With this new response handling, it seems there's some information getting lost.

If you tweak the `get_tx_by_hash_success` test to have a bogus hash (just arbitrarily change one of the digits), the server responds with just `{"jsonrpc":"2.0","id":0,"result":null}`.

The way I had it before, that response was conveyed as: `failed to get transaction by hash: InterpretationError { msg: "invalid type: null, expected struct Transaction at line 1 column 37", request_body: "{\"version\":\"2.0\",\"method\":\"eth_getTransactionByHash\",\"params\":[\"0xf008e9f9bb92057dd0035496fbf4fb54f66b4b18b370928e46d6603933054d5a\"],\"id\":0}", expected_type: "rethnet_eth::remote::jsonrpc::Success<rethnet_eth::remote::eth::Transaction>", response_text: "{\"jsonrpc\":\"2.0\",\"id\":0,\"result\":null}" }', crates/rethnet_eth/src/remote.rs:516:14`

With the revised error handling, now you just get: ```thread 'remote::client::tests::get_tx_by_hash_success' panicked at 'Response `{"jsonrpc":"2.0","id":0,"result":null}` failed to parse with expected type `rethnet_eth::remote::eth::Transaction`, due to error: data did not match any variant of untagged enum ResponseData at line 1 column 38', crates/rethnet_eth/src/remote/client.rs:73:17```

With the new way, we can't see what the request was, nor where it originated (in the new way, the panic originates from the internal function rather than the site of usage (eg the test)). I think it would be good to restore those two pieces of information to the error message.

Reply:
I added a test just to be clear what i mean. it seems to me that the case exercised by that test should be a recoverable error rather than a `panic`. let me know what you think.